src/lfortran/semantics/ast_common_visitor.h

```diff
@@ -2468,6 +2461,9 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                 type = ASRUtils::TYPE(ASR::make_CPtr_t(al, loc));
             } else {
                 if (!v) {
+                    if (is_template) {
+                      throw SemanticError("Type parameter '" + derived_type_name + "' not required", loc);
```

Comment:
Something like this I think is better:
```suggestion
                      throw SemanticError("Type parameter '" + derived_type_name + "' is not part of any requirement", loc);
```